### DISCUSSION TITLE
Should route controller take into account `ListenerStatus.SupportedKinds`?

### DISCUSSION THREAD

**Discussion by pmalek** [kubernetes-sigs/gateway-api#1555] [Q&A]
Given https://github.com/kubernetes-sigs/gateway-api/blob/a8f72f5778d9d769dfda5d4332d23529aeb7f2e1/apis/v1beta1/gateway_types.go#L656-L666 should an implementation of route controller take into account `ListenerStatus.SupportedKinds`?

It seems to me that this is purely informational.

If, in some case this should be taken into account then what reason should be used when the route kind is not in the supported kinds?

**Comment by howardjohn** [comment]
There is only 1 controller - which controls Gateway and Route. I don't think there can be a split brain controller (?). Given this, the thing programming routes and writing the status are the same thing.

The controller should take this into account, although literally reading from Status may not be the best way to implement it.

**Reply by pmalek** [reply]
Maybe I mentioned the controller unnecessarily (I meant here the controller that might be a part of one's operator: you can have controller that only watches and reconciles Gateways, you can have one that reconciles HTTPRoutes etc.).

> although literally reading from Status may not be the best way to implement it.

Can you elaborate on this though? What other way(s) do you see?

**Reply by youngnick** [reply]
I think that, because of the complexity of having multiple separate controllers that need to know about updates to their relevant objects - particularly the fact that updates to Gateway can cause Routes to need to be re-reconciled, most implementations are using a single controllers that watches *all* of the Gateway resources and performs a full reconciliation whenever *any* of them are updated.

In general though, reading from Status to trigger other things to be reconciled is not a great idea because it's very racy. It's *far* easier to prevent races between different components when you have everything in the one instead. 😄

**Reply by shaneutt** [reply]
In this comment thread so far it seems there's some assumption being made that all implementations should have the same controller which is enforcing the policy of whether or not a kind is supported also be the one determining that policy (in which case the policy is, as @pmalek was asking, basically just "for show" or otherwise for the user's benefit). I'm having a hard time convincing myself that this is a fair assumption to make: it would seem reasonable to me that there be architectures where the controller responsible for managing the _lifecycle_ of a `Gateway` is distinct from the controller responsible for _pushing configuration_ to that `Gateway` according to routing information like `HTTPRoute`. It seems to me that such a distinction would help to avoid monolithic building and help provide better separation of concerns, no?

**Reply by pmalek** [reply]
> It seems to me that such a distinction would help to avoid monolithic building and help provide better separation of concerns, no?

It would but in my view it's totally up to the developers of particular implementations. Gateway API spec can encourage people to split those controllers - at the very least for reasons that @shaneutt  mentioned - by explicitly mentioning this somewhere but it'd be 100% up to designers/developers of particular implementations which path they choose.

I do see @youngnick's point as well about this being potentially racy, which is true but OTOH in k8s eventual consistency world, we shouldn't really rely on the order of things happening, should we? The end results given a particular configuration/resource set should be the same regardless, right? I might be missing some use cases which would be prone to those though.

**Reply by shaneutt** [reply]
I think I understand both John and Nick's hesitation in this specific case (and I share it some), treating the `SupportedKinds` as an effective **policy declaration** adds considerable complexity: it goes beyond just enqueuing child objects when the parent changes, those enqueued reconciliations also need to reference back to the field which changed and triggered the event in the first place. I can see the reference-back to the `Gateway` by the enqueued routes being "racy" as Nick suggested in that if multiple changes happen rapidly your dataplane configuration may do a bit of an odd dance (dropping/adding routes). However I also agree that this can be made eventually consistent as Patryk suggested.

My own experience has taught me the same lesson John and Nick have expressed: trying to use status to trigger reconciliation can be messy and I would avoid it when possible. I think it's worth keeping in mind however that sometimes there's good reasons (e.g. `Deployment` controller). As for prescribing the use of a single controller however to avoid this complexity: I feel _strongly_ that we can't readily discourage multi-controller implementations as there are potential architectural reasons and benefits of these. It also occurs to me that what we're really trying to tackle here is "how to react to changes in `SupportedKinds` throughout the lifecycle of a `Gateway` resource", as the insinuation of this field being present in `Gateway` status is that it may change throughout the lifecycle of the `Gateway`, and to that end I can see why there may be an inclination to respond to the status.

I guess I pose the question to all of you: @pmalek @howardjohn and @youngnick: Is "how to react to changes in `SupportedKinds` throughout the lifecycle of a `Gateway` resource" the problem we're trying to solve and we're on the same page? I was getting a sense from Nick at least that he might have worked on implementations which consider this entirely static information, so I wanted to check.

**Reply by youngnick** [reply]
Good discussion, thanks everyone.

I think that what I want to say is something like the official position on Conditions: They're made primarily for humans, rather than for machine consumption. That doesn't mean you _can't_ or _shouldn't_ use them, just that you should be _careful_ if you do.

Nothing I've worked on assumes that they'll be static, but I just haven't ever keyed anything off changes, because of the raciness concerns I listed.

**Reply by pmalek** [reply]
> how to react to changes in SupportedKinds throughout the lifecycle of a Gateway resource" the problem we're trying to solve and we're on the same page

Partially yes. I'd say I'm also not sure how to react to them in general (see below).

I'm not 100% convinced what to do in situations like e.g.: user defined `Listener.AllowedRoutes`, the implementation posts its `SupportedKinds` in `ListenerStatus` and it cannot support at least one of the route type from the specified `Listener.AllowedRoutes`. Should those be subsequently allowed or not? Am I the only one seeing this discrepancy or am I the only one missing something 😅 ?

**Reply by youngnick** [reply]
Ah, I thought that only the union of those two sets of things would be valid - if a user _allows_ routes that an implementation doesn't support, then it's expected that the implementation would not accept those routes. (It can't support them, anyway!)

In that case, the SupportedKinds is intended as a hint for the user to see what Routes _are_ supported, but the implementation also should record that the unsupported Route's attachment to its parentRef failed because the Kind is not supported by that Gateway implementation. Does that make sense?

So, the SupportedKinds is more a _hint_ for the user about what the implementation _is capable of doing_, not a request for anything. (In general, `status` is intended to record the response from "the system" (being Kubernetes and all the installed controllers) about what's been requested in the `spec`.

It does seem like this could use clarification though.

**Reply by pmalek** [reply]
Got it. So basically instead of checking the `SupportedKinds` we should "defer" to the implementation to reject that.

Maybe the fact I was even questioning this approach is that I'm biased by Kong's implementation which has several layers of verification whether e.g. a route could be attached and instead of invalidating it at the gateway level and relying on error received from it (which would be hard to then untangle and write status updates for) we opted in (for now) to look at the status and have that information readily available at that point (where we now for each entities we failed to attach).

Does that make sense? 🤔

**Reply by youngnick** [reply]
I think so, I guess I've always assumed that an implementation will know what Kinds it supports in itself, so you shouldn't need to check the SupportedKinds. But if the same thing is setting the status and reading it, I guess it's fine (if a little odd).
